src/tls/socket-activation-helper.c

```diff
@@ -0,0 +1,193 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+#define _GNU_SOURCE 1
+
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/poll.h>
+#include <sys/socket.h>
+#include <sys/wait.h>
+#include <stdlib.h>
+#include <sys/un.h>
+#include <errno.h>
+#include <err.h>
+
+#define N_ELEMENTS(arr) (sizeof (arr) / sizeof ((arr)[0]))
+#define SD_LISTEN_FDS_START 3
+#define MAX_COCKPIT_WS_ARGS 5
+
+static struct instance_type
+{
+  const char *sockname;
+  const char *argv[MAX_COCKPIT_WS_ARGS];
+} instance_types[] = {
+  {"https.sock", {"--for-tls-proxy", "--port", "0"}},
+  {"http-redirect.sock", {"--proxy-tls-redirect", "--no-tls", "--port", "0"}},
+  {"http.sock", {"--no-tls", "--port", "0"}},
+};
+
+static int socket_to_pid[N_ELEMENTS (instance_types)];
+static struct pollfd ufds[N_ELEMENTS (instance_types)];
+
+static void
+handle_cockpit_ws_exited (int sigid)
+{
+  pid_t pid;
+  while ((pid = waitpid (-1, NULL, WNOHANG)) > 0)
+    {
+      for (int i = 0; i < N_ELEMENTS (instance_types); i++)
+        {
+          if (pid == socket_to_pid[i])
+            {
+              ufds[i].events = POLLIN;
+              socket_to_pid[i] = 0;
+              return;
+            }
+        }
+    }
+  /* This can only be reached in case cockpit-ws exited fast and the parent
+   * process did not manage to insert its PID into the array
+   */
+  errx (1, "Could not find the process in socket_to_pid array");
+}
+
+static pid_t
+spawn_cockpit_ws (const char *ws_path, int fd,
+                  struct instance_type sock_instance)
+{
+  pid_t pid = fork ();
+
+  if (pid == -1)
+    {
+      err (1, "fork() failed");
+    }
+  else if (pid > 0)
+    {
+      return pid;
+    }
+  else
+    {
+      const char *args[MAX_COCKPIT_WS_ARGS + 1];
+      int i = 0;
+      static char listen_pids[20];
+
+      /* pass the socket to ws like systemd activation does, see sd_listen_fds(3) */
+      if (dup2 (fd, SD_LISTEN_FDS_START) < 0)
+        err (1, "dup2() failed");
+
+      if (snprintf (listen_pids, sizeof (listen_pids), "LISTEN_PIDS=%i",
+                    getpid ()) >= sizeof listen_pids)
+        errx (1, "asprintf() failed");
+
+      char *env[] = {
+        "LISTEN_FDS=1", listen_pids, "SD_LISTEN_FDS_START=3", (char *) 0
```

Comment:
Putting a string literal into a `char *` ought to be triggering a warning.  We apparently don't have `-Wwrite-strings` enabled (surprisingly).

`SD_LISTEN_FDS_START` is not anywhere specified as an environment variable that should be set, but `LISTEN_FDNAMES` is, and it's missing.

Finally, you can't just list these few items in the `envp` that you pass to `execve()` because you also need to make sure the common standard variables like `PATH`, `HOME`, etc. will continue to be there.  The old `wsinstance.c` code used `setenv()` after `fork()` and then `execl()` instead of `execve()`.  `execl()` won't work for you, but its non-environment-modifying equivalent `execv()` is probably exactly what you're looking for here.

Overall, my recommendation here is to take a look at the way that `wsinstance.c` is doing it and follow it as closely as possible (adjusting for the fact that your arguments are in an array rather than hardcoded).